Protobuf Local Reference support

Jtalk opened the pull request:
Adding support for Protobuf Local References. 

We're using [Wire](https://github.com/square/wire) for Protobuf manipulations. 

I've included a unit-test suite for proto3, but nothing for proto2, sorry. 

Resolves #77

ImFlog commented on src/main/kotlin/com/github/imflog/schema/registry/parser/ProtobufSchemaParser.kt:
```diff
@@ -2,21 +2,300 @@ package com.github.imflog.schema.registry.parser
+                // so it'll never match unless we strip the dot).
+                .removePrefix(".")
+            val fullType = source.typesAndNestedTypes().find { it.type.toString().endsWith(withoutDot) }
+                ?: throw RuntimeException("Type $type could not be found in ${source.location.path}")
```
Marker about the exception typing.

ImFlog commented on src/main/kotlin/com/github/imflog/schema/registry/parser/ProtobufSchemaParser.kt:
```diff
@@ -2,21 +2,300 @@ package com.github.imflog.schema.registry.parser
+
+        @Suppress("UNUSED_PARAMETER")
+        private fun standardizeNames(service: Service, file: ProtoFile): Service {
+            // RPC rewriting is not supported. RPCs are not fully parsed by Wire, so getting them
+            // supported would be a lot of work.
```
Should we also log a warn about this for end-user ?

ImFlog commented on src/main/kotlin/com/github/imflog/schema/registry/parser/ProtobufSchemaParser.kt:
```diff
@@ -2,21 +2,300 @@ package com.github.imflog.schema.registry.parser
+
+        @Suppress("UNUSED_PARAMETER")
+        private fun standardizeNames(extensions: Extensions, file: ProtoFile): Extensions {
+            return extensions // Extension overwrite is not supported
```
Should we also log a warn about this for end-user ?

ImFlog reviewed (commented):
Hello @Jtalk tank you very much for this great PR!

As I am not a big user of protobuf myself It feels great to see extensive Unit Test on the feature !
Two things (apart from the review in itself): 
1. Can you update the integration test that references the local support to add a case with Protobuf (commented today I think) ?
2. Can you update the example directory to handle this too ? Even if it's not released yet It would be easier for me for the release to do a last minute check and it provides support for the users.

Thank you again 🙇

Jtalk commented on src/main/kotlin/com/github/imflog/schema/registry/LocalReference.kt:
```diff
@@ -4,7 +4,7 @@ import java.io.File
     fun content(rootDir: File) = rootDir.resolve(path).readText()
-}
\ No newline at end of file
+}
```
It's still in use:

![image](https://github.com/ImFlog/schema-registry-plugin/assets/1573833/4bbcb5f6-24cc-4a88-b452-80bd3fb3f9d0)

Jtalk commented on src/main/kotlin/com/github/imflog/schema/registry/parser/ProtobufSchemaParser.kt:
```diff
@@ -2,21 +2,300 @@ package com.github.imflog.schema.registry.parser
+        }
+
+        override fun next(): ProtoFile {
+            val current = imports.removeFirstOrNull() ?: throw NoSuchElementException("No more imports")
```
In general yes, I'll update the PR. 

For this specific line, we shouldn't. `next()` is implementing `Iterator.next()`, and throwing anything else for a missing record would make it an invalid implementation:

![image](https://github.com/ImFlog/schema-registry-plugin/assets/1573833/cc9e75c7-d12b-4893-b029-78873c5647d8)

 On the bright side, if the iterator is used correctly, the end user would not be able to ever see this exception.

Jtalk commented on src/main/kotlin/com/github/imflog/schema/registry/parser/ProtobufSchemaParser.kt:
```diff
@@ -2,21 +2,300 @@ package com.github.imflog.schema.registry.parser
+        }
+
+        override fun next(): ProtoFile {
+            return if (!rootReturned) {
+                rootReturned = true
+                root
+            } else {
+                delegate.next()
+            }
```
And on the next call `if !rootReturned` is `false`, so we'll call the `delegate` instead.

I've added kdocs to both iterators with some high level explanation, hopefully it'll make it clearer.

Jtalk commented on src/main/kotlin/com/github/imflog/schema/registry/parser/ProtobufSchemaParser.kt:
```diff
@@ -2,21 +2,300 @@ package com.github.imflog.schema.registry.parser
     ): String {
-        throw NotImplementedError("Local reference is not available for Protobuf yet")
+        val schema = schemaFor(rootDir)
+        val source = schema.protoFile(File(schemaPath).relativeTo(rootDir).path)!!
```
True, good spot

Jtalk commented:
Hey @ImFlog, I've fixed/responded to your comments now. 

I've added an example, but it won't compile due to how Wire is configured right now: it links the whole Protobuf model, meaning it's not going to work with 2 files providing the same message (Company) sitting within the same Protobuf root. It's possible to either skip linking & doing import lookup ourselves, based on the root dir, or to make the root dir configurable (something I wanted to ask for separately, as our schema files live outside the Gradle project the plugin is in). 

I think it's in a useful state right now, but I have another deliverable coming up, so I won't be much help with this for the next few weeks. We could either merge it as is, or keep it open for a bit & I'll revisit it later this year.

ImFlog commented:
Hello,

To overcome the issue in the example folder could we use a specific protobuf directory for this case ?
Not sure to understand correctly, another thing could be to create a new project in the example root project ?

Overall I think it's better to be able to merge this and create a new issue for this whole directory path thing for later reference, I'll be glad to have a new PR from you in the future :)

ImFlog commented:
Hello @Jtalk,
Do you have by any chance a bit of time to finish this MR ?
It's very close to be merged by I think you forgot to push the fixes.
Thank you for the contribution 🙏

Jtalk commented:
Hey @ImFlog, my apologies, I have indeed forgotten to push the fixes. 

It's done now. I've also rebased the PR & run the tests again, it all seems fine.

I can still make small adjustments, but overhauling the Protobuf examples would take longer than I can commit right now, I'm afraid. I should have some time in 1-2 months, but I agree, it'd be better to merge it now & revisit the path issues later. The Protobuf flow is working, it's mostly the limitation of the existing `examples/` structure that's holding us back here.

ImFlog commented:
I've reverted the last commit that made the example folder fail and created an issue for the configurable folder directory.
I'm merging this and release a new version.
